homeassistant/helpers/config_entry_flow.py

```diff
@@ -81,7 +81,6 @@ async def async_step_discovery(
 
         return await self.async_step_confirm()
 
-    async_step_zeroconf = async_step_discovery
```

Comment:
This line is redundant with the base class, and actually causing `mypy` to crash, so had to remove it!!! See linked PR (#59576)
https://github.com/home-assistant/core/blob/363de374003221bc3a953bc049471ea237ef0498/homeassistant/config_entries.py#L1371-L1375

Reply:
Did you open in issue with mypy?

Reply:
I have not (yet) opened the issue with mypy, because I thought the line was just redundant. I will try to open an issue with them next week.
@balloob suggests that the line needs to stay there - so maybe we have to keep it and set it to ignore the type?

Or maybe there is another way to set a different signature for `async_step_zeroconf` (with `ZeroconfServiceInfo`) and `async_step_discovery` (`DiscoveryInfoType`) ?

Reply:
Ignore is fine here. The alternative would be to write out the function and await the discovery handler which would add another await only to satisfy mypy which doesn't seem like a good trade off to me.

Reply:
I didn't manage to make `#type: ignore` work.

If I add the line back in, the I have to remove altogether the type hint in `sonos` component (it seems to be the only component using `DiscoveryFlowHandler` with zeroconf and type hints).

Is removing the type hint on `sonos` a good (at least temporary) trade off?

Reply:
We can just make these simple proxy methods. A method with signature calling the async_step_discovery. Will retain typing, will retain functionality and fixes the issue?

Reply:
@frenck I have update the code as per your suggestion. @bdraco didn't think it was a good trade off though.

Reply:
I don't have strong feelings either way.  I tend to be more a bit more nervous about adding `await`s to discovery since we still have a race conditions I haven't been able to sort out that result in duplicate discovery.

Reply:
FYI: I have managed to get a smaller sample class to crash mypy, so I opened an issue there.
See: https://github.com/python/mypy/issues/11556

Reply:
This issue is still present in mypy `0.920+dev.053a1beb94ee4e5b3260725594315d1b6776e42f`, and that prevents us from using the `# type: ignore`.
I think it would be a shame not to use type hints on Zeroconf discover flows.

Reply:
@bdraco I have removed the `await` keyword, and it now returns the generic `async_step_discovery` coroutine directly.
Does that work for you?

Reply:
If I remove the await keyword, now `pylint` complains with `invalid-overridden-method`...
I have reverted the change, but I could bring it back in with pylint disable:
```python
    def async_step_zeroconf(  # pylint: disable=invalid-overridden-method
        self,
        discovery_info: zeroconf.ZeroconfServiceInfo,
    ) -> Coroutine[Any, Any, FlowResult]:
        """Handle a flow initialized by Zeroconf discovery."""
        return self.async_step_discovery(cast(dict, discovery_info))
```

Reply:
last, but also ugly, but probably the best solution is code duplication:

```py
    async def async_step_zeroconf(
        self, discovery_info: zeroconf.ZeroconfServiceInfo
    ) -> FlowResult:
        """Handle a flow initialized by Zeroconf discovery."""
        if self._async_in_progress() or self._async_current_entries():
            return self.async_abort(reason="single_instance_allowed")

        await self.async_set_unique_id(self._domain)

        return await self.async_step_confirm()
```

Reply:
Happy to implement. @bdraco do you agree?

Reply:
I agree 👍

Reply:
Great. I have updated the code. Hopefully the CI passes and we are good to go...